include/spblas/vendor/rocsparse/multiply_spgemm.hpp

```diff
@@ -0,0 +1,326 @@
+        0, c.rowptr().data(), NULL, NULL,
+        to_rocsparse_indextype<typename output_type::offset_type>(),
+        to_rocsparse_indextype<typename output_type::index_type>(),
+        rocsparse_index_base_zero,
+        to_rocsparse_datatype<typename output_type::scalar_type>()));
+    __rocsparse::throw_if_error(rocsparse_create_csr_descr(
+        &this->mat_d_, __backend::shape(d_base)[0], __backend::shape(d_base)[1],
+        d_base.values().size(), d_base.rowptr().data(), d_base.colind().data(),
+        d_base.values().data(),
+        to_rocsparse_indextype<typename output_type::offset_type>(),
+        to_rocsparse_indextype<typename output_type::index_type>(),
+        rocsparse_index_base_zero,
+        to_rocsparse_datatype<typename output_type::scalar_type>()));
+    //--------------------------------------------------------------------------
+    // SpGEMM Computation
+    // ask buffer_size bytes for external memory
+    __rocsparse::throw_if_error(rocsparse_spgemm(
+        handle, rocsparse_operation_none, rocsparse_operation_none, &alpha,
+        this->mat_a_, this->mat_b_, &beta, this->mat_d_, this->mat_c_,
+        to_rocsparse_datatype<value_type>(), rocsparse_spgemm_alg_default,
+        rocsparse_spgemm_stage_buffer_size, &buffer_size, nullptr));
+    if (buffer_size > this->buffer_size_) {
+      this->alloc_.deallocate(workspace_, this->buffer_size_);
+      this->buffer_size_ = buffer_size;
+      workspace_ = this->alloc_.allocate(buffer_size);
+    }
+    __rocsparse::throw_if_error(rocsparse_spgemm(
+        handle, rocsparse_operation_none, rocsparse_operation_none, &alpha,
+        this->mat_a_, this->mat_b_, &beta, this->mat_d_, this->mat_c_,
+        to_rocsparse_datatype<value_type>(), rocsparse_spgemm_alg_default,
+        rocsparse_spgemm_stage_nnz, &this->buffer_size_, this->workspace_));
+    // get matrix C non-zero entries and size
+    int64_t c_num_rows;
+    int64_t c_num_cols;
+    __rocsparse::throw_if_error(rocsparse_spmat_get_size(
+        this->mat_c_, &c_num_rows, &c_num_cols, &this->result_nnz_));
+    this->result_shape_ = index<index_t>(c_num_rows, c_num_cols);
+  }
+
+  template <matrix A, matrix B, matrix C, matrix D>
+    requires __detail::has_csr_base<A> && __detail::has_csr_base<B> &&
+             __detail::is_csr_view_v<C> && __detail::has_csr_base<D>
+  void multiply_fill(A&& a, B&& b, C&& c, D&& d) {
+    auto a_base = __detail::get_ultimate_base(a);
+    auto b_base = __detail::get_ultimate_base(b);
+    using matrix_type = decltype(a_base);
+    using input_type = decltype(b_base);
+    using output_type = std::remove_reference_t<decltype(c)>;
+    using value_type = typename matrix_type::scalar_type;
+
+    auto alpha_optional = __detail::get_scaling_factor(a, b);
+    tensor_scalar_t<A> alpha = alpha_optional.value_or(1);
+    value_type alpha_val = alpha;
+    auto beta_optional = __detail::get_scaling_factor(d);
+    value_type beta = beta_optional.value_or(1);
```

Comment:
I am not sure that relying on a and d types for alpha and beta is the best approach.